tools/newgtlds.go

```diff
@@ -0,0 +1,254 @@
+// newgtlds is a utility command that downloads the list of gTLDs from ICANN
+// and formats it into the PSL format, writing to stdout.
+package main
+
+import (
+	"bytes"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"net/http"
+	"os"
+	"strings"
+	"text/template"
+	"time"
+)
+
+// ICANN_GTLD_JSON_URL is the URL for the ICANN gTLD JSON registry (version
+// 2). See https://www.icann.org/resources/pages/registries/registries-en for
+// more information.
+const ICANN_GTLD_JSON_URL = "https://www.icann.org/resources/registries/gtlds/v2/gtlds.json"
+
+var (
+	// legacyGTLDs are gTLDs that predate ICANN's new gTLD program. These legacy
+	// gTLDs are present in the ICANN_GTLD_JSON_URL data but we do not want to
+	// include them in the new gTLD section of the PSL data because it will create
+	// duplicates with existing entries alongside registry-reserved second level
+	// domains present in the PSL data. Entries present in legacyGTLDs will not be
+	// output by this tool when generating the new gTLD data.
+	legacyGTLDs = map[string]bool{
+		"aero":   true,
+		"asia":   true,
+		"biz":    true,
+		"cat":    true,
+		"com":    true,
+		"coop":   true,
+		"info":   true,
+		"jobs":   true,
+		"mobi":   true,
+		"museum": true,
+		"name":   true,
+		"net":    true,
+		"org":    true,
+		"post":   true,
+		"pro":    true,
+		"tel":    true,
+		"xxx":    true,
+	}
+
+	// pslTemplate is a parsed text/template instance for rendering a list of pslEntry
+	// objects in the format used by the public suffix list.
+	//
+	// It expects the following template data:
+	//   URL - the string URL that the data was fetched from.
+	//   Date - the time.Date that the data was fetched.
+	//   Entries - a list of pslEntry objects.
+	pslTemplate = template.Must(template.New("public-suffix-list-gtlds").Parse(`
+// List of new gTLDs imported from {{ .URL }} on {{ .Date.Format "2006-01-02T15:04:05Z07:00" }}
+// This list is auto-generated, don't edit it manually.
+
+{{- range .Entries }}
+{{ .Comment }}
+{{ printf "%s\n" .ULabel}}
+{{- end }}
+`))
+)
+
+// pslEntry is a struct matching a subset of the gTLD data fields present in
+// each object entry of the "GLTDs" array from ICANN_GTLD_JSON_URL.
+type pslEntry struct {
+	// GTLD contains the gTLD name. For internationalized gTLDs the GTLD field is
+	// expressed in punycode.
+	GTLD string
```

Comment:
nit: For consistency with ULabel, should this be called ALabel? I'm fine with either, but the latter seems slightly more descriptive of format - but at the cost of being less-descriptive of content.

Reply:
The consistency argument makes sense to me. Implemented in ae671bb